modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+        end
+      end
+    end
+
+    [flag_allowed_to_enroll, allowed_sids]
+  end
+
+  def query_ldap_server(raw_filter, attributes, base_prefix: nil)
+    ldap_connect do |ldap|
+      validate_bind_success!(ldap)
+
+      if !@base_dn.blank?
+        vprint_status("Using already discovered base DN: #{@base_dn}")
+      elsif (@base_dn = datastore['BASE_DN'])
+        print_status("User-specified base DN: #{@base_dn}")
+      else
+        print_status('Discovering base DN automatically')
+
+        unless (@base_dn = discover_base_dn(ldap))
+          print_warning("Couldn't discover base DN!")
+        end
+      end
+
+      if @base_dn.blank?
+        fail_with(Failure::BadConfig, 'No base DN was found or specified, cannot continue!')
+      end
+
+      if base_prefix.blank?
+        full_base_dn = @base_dn.to_s
+      else
+        full_base_dn = "#{base_prefix},#{@base_dn}"
+      end
+      begin
+        filter = Net::LDAP::Filter.construct(raw_filter)
+      rescue StandardError => e
+        fail_with(Failure::BadConfig, "Could not compile the filter to find the ESC1 objects! Error was #{e}")
+      end
+
+      returned_entries = ldap.search(base: full_base_dn, filter: filter, attributes: attributes)
+      query_result = ldap.as_json['result']['ldap_result']
+
+      result_code, result_message = check_query_result_code(query_result, filter)
+      case result_code
+      when -1
+        fail_with(Failure::BadConfig, result_message)
+      when 0
+        vprint_good(result_message)
+      when 1
+        fail_with(Failure::NoAccess, result_message)
+      when 2
+        fail_with(Failure::UnexpectedReply, result_message)
+      end
+
+      if returned_entries.blank?
+        print_error("No results found for #{filter}.")
+
+        nil
+      else
+
+        returned_entries
+      end
+    end
+  rescue Rex::ConnectionTimeout
+    fail_with(Failure::Unreachable, "Couldn't reach #{datastore['RHOST']}!")
+  rescue Net::LDAP::Error => e
+    fail_with(Failure::UnexpectedReply, "Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+  end
+
+  def convert_sids_to_human_readable_name(sids_array)
+    output = []
+    for sid in sids_array
+      raw_filter = "(objectSID=#{sid})"
+      attributes = ['dn', 'description']
+      sid_entry = query_ldap_server(raw_filter, attributes)
+      fail_with(Failure::UnexpectedReply, 'Could not find any matching SID!') if sid_entry.blank?
+      if sid_entry[0][:description]
+        output << [sid, sid_entry[0][:dn][0], sid_entry[0][:description][0]]
+      else
+        output << [sid, sid_entry[0][:dn][0], nil]
+      end
+    end
+
+    result = ''
+    output.each do |sid_string, dn, description|
+      if description
+        result += "#{sid_string} aka #{dn} (#{description}) | "
+      else
+        result += "#{sid_string} aka #{dn} | "
+      end
+    end
+
+    result.gsub(%r{ \| $}, '') # Strip out the last trailing " \ " from the output.
```

Comment:
You should just construct an array and join it instead of applying a regex to remove the trailing delimiter.

Reply:
I'm unclear on how this would be done without loosing out on the "aka" part of the string. Would you be able to clarify this a little bit as I think its a good idea, just struggling to picture how to do it without loosing wording, which could very well be the wrong approach to take here.

Reply:
Build an array of strings, the content of the array can be whatever including your AKA part. The important thing is that you're not preemptively adding the ` | ` delimiter suffix.

```
    result = []
    output.each do |sid_string, dn, description|
      if description
        result << "#{sid_string} aka #{dn} (#{description})"
      else
        result << "#{sid_string} aka #{dn}"
      end
    end
    result.join(' | ')
```

Reply:
Thanks, much appreciated, will implement this now 👍